lib/services/local/access_monitoring_rules.go

```diff
@@ -110,3 +113,58 @@ func (s *AccessMonitoringRulesService) DeleteAccessMonitoringRule(ctx context.Co
 func (s *AccessMonitoringRulesService) DeleteAllAccessMonitoringRules(ctx context.Context) error {
 	return trace.Wrap(s.svc.DeleteAllResources(ctx))
 }
+
+// ListAccessMonitoringRulesWithFilter returns a paginated list of access monitoring rules that match the given filter.
+func (s *AccessMonitoringRulesService) ListAccessMonitoringRulesWithFilter(ctx context.Context, pageSize int, pageToken string, subjects []string, notificationName string) ([]*accessmonitoringrulesv1.AccessMonitoringRule, string, error) {
+
+	rangeStart := backend.Key(accessMonitoringRulesPrefix, pageToken)
+	rangeEnd := backend.RangeEnd(backend.ExactKey(accessMonitoringRulesPrefix))
+
+	// Get most limit+1 results to determine if there will be a next key.
+	maxLimit := pageSize + 1
+	var resources []*accessmonitoringrulesv1.AccessMonitoringRule
+	if err := backend.IterateRange(ctx, s.backend, rangeStart, rangeEnd, maxLimit, func(items []backend.Item) (stop bool, err error) {
+		for _, item := range items {
+			if len(resources) == maxLimit {
+				break
+			}
+
+			accessMonitoringRule, err := services.UnmarshalAccessMonitoringRule(item.Value)
+			if err != nil {
+				return false, trace.Wrap(err)
+			}
+			if ok := match(accessMonitoringRule, subjects, notificationName); ok {
+				resources = append(resources, accessMonitoringRule)
+			}
+		}
+
+		return len(resources) == maxLimit, nil
+	}); err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	var nextKey string
+	if len(resources) > pageSize {
+		nextKey = resources[len(resources)-1].Metadata.Name
+		// Truncate the last item that was used to determine next row existence.
+		resources = resources[:pageSize]
+	}
+
+	return resources, nextKey, nil
+}
+
+func match(rule *accessmonitoringrulesv1.AccessMonitoringRule, subjects []string, notificationName string) bool {
+	for _, subject := range subjects {
+		if ok := slices.ContainsFunc(rule.Spec.Subjects, func(s string) bool {
+			return s == subject
+		}); !ok {
+			return false
+		}
+	}
+	if notificationName != "" {
```

Comment:
Can we do the notification name check first? It might fail faster than the doubly-nested for loop above.